libdnf/module/ModulePackage.cpp

```diff
@@ -0,0 +1,181 @@
+                    } else {
+                        hasStreamRequires = true;
+                        ss << ":" << moduleStream << ")";
+                        depId = pool_str2id(pool, ss.str().c_str(), 1);
+                        solvable_add_deparray(solvable, SOLVABLE_REQUIRES, depId, 0);
+                    }
+                }
+
+                if (!hasStreamRequires) {
+                    ss.str(std::string());
+                    ss << "module-n(" + moduleName + ")";
+                    depId = pool_str2id(pool, ss.str().c_str(), 1);
+                    solvable_add_deparray(solvable, SOLVABLE_REQUIRES, depId, -1);
+                }
+            }
+        }
+    }
+}
+
+std::string ModulePackage::getName() const
+{
+    return metadata->getName();
+}
+
+std::string ModulePackage::getNameStream() const
+{
+    std::ostringstream ss;
+    ss << getName() << ":" << getStream();
+    return ss.str();
+}
+
+std::string ModulePackage::getNameStreamVersion() const
+{
+    std::ostringstream ss;
+    ss << getNameStream() << ":" << getVersion();
+    return ss.str();
+}
+
+std::string ModulePackage::getStream() const
+{
+    return metadata->getStream();
+}
+
+std::string ModulePackage::getVersion() const
+{
+    return std::to_string(metadata->getVersion());
+}
+
+std::string ModulePackage::getContext() const
+{
+    return metadata->getContext();
+}
+
+std::string ModulePackage::getArchitecture() const
+{
+    return metadata->getArchitecture();
+}
+
+std::string ModulePackage::getFullIdentifier() const
+{
+    std::ostringstream ss;
+    ss << getName() << ":" << getStream() << ":" << getVersion() << ":" << getContext() << ":" << getArchitecture();
+    return ss.str();
+}
+
+std::string ModulePackage::getSummary() const
+{
+    return metadata->getSummary();
+}
+
+std::string ModulePackage::getDescription() const
+{
+    return metadata->getDescription();
+}
+
+std::vector<std::string> ModulePackage::getArtifacts() const
+{
+    return metadata->getArtifacts();
+}
+
+std::shared_ptr<Profile> ModulePackage::getProfile(const std::string &name) const
+{
+    return metadata->getProfile(name);
+}
+
+std::vector<std::shared_ptr<ModuleProfile> > ModulePackage::getProfiles() const
+{
+    return metadata->getProfiles();
+}
+
+std::vector<std::shared_ptr<ModuleDependencies> > ModulePackage::getModuleDependencies() const
+{
+    return metadata->getDependencies();
+}
+
+bool ModulePackage::isEnabled()
+{
+    return state->isEnabled();
+}
+
+void ModulePackage::setState(const std::shared_ptr<ModuleState> &state)
+{
+    this->state = state;
+}
+
+void ModulePackage::enable()
+{
+    state = std::make_shared<EnabledState>();
+}
+
+void ModulePackage::addStreamConflict(const std::shared_ptr<ModulePackage> &package)
+{
+    std::ostringstream ss;
+    Solvable *solvable = pool_id2solvable(pool, id);
+
+    ss << "module-ns(" + package->getNameStream() + ")";
```

Comment:
no, just `module(n)`, modules conflict by name and not by stream.

Reply:
This makes a module conflicting with other existing streams.
We're "installing" all module versions in a stream at once.
Conflicting by module name would cause these versions conflicting with each other.
Or am I missing something?

Reply:
> This makes a module conflicting with other existing streams.

No, it makes module conflicting with other versions in the same stream. What I proposed is exactly what you described in your line.

> We're "installing" all module versions in a stream at once.
> Conflicting by module name would cause these versions conflicting with each other.

That's what you have right now already, because a:1:1 | a:1:2 provide a:1 and you conflict with a:1.

> Or am I missing something?

You do.